### PR TITLE
[thread-cert] add test to verify MED can inform previous parent on reattach

### PR DESCRIPTION
**Pull request by simonlingoogle**
This commit verifies that MEDs can inform previous parent on reattach with `OPENTHREAD_CONFIG_MLE_INFORM_PREVIOUS_PARENT_ON_REATTACH` enabled.

### CODE REVIEW

File: tests/scripts/thread-cert/test_inform_previous_parent_on_reattach.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            'is_mtd': True,
+            'mode': 'rn',
+            'allowlist': [LEADER],
+            'timeout': LONG_CHILD_TIMEOUT,
+        },
+    }
+
+    def test(self):
+        if 'posix' in os.getenv('OT_CLI_PATH', ''):
+            self.skipTest("skip for posix tests")
```
Comment (abtink) [inline tests/scripts/thread-cert/test_inform_previous_parent_on_reattach.py:81]: What is the reason for skipping under `posix`?

**Review by abtink — APPROVED** [review]
Thanks @simonlingoogle for adding. Looks great. 👍 

I think we a test-case similar to this under `toranj` but I agree (and I think would be good) to re-add/move (some of) the test-cases under `thread-cert` (and use CLI).

File: tests/scripts/thread-cert/node.py [hunk 2]
```diff
@@ -1572,6 +1572,71 @@ def remove_service(self, enterpriseNumber, serviceData):
...
+            }
+
+            if 'Ver' in headers:
+                child['ver'] = int(col('Ver'))
+
+            if 'CSL' in headers:
+                child['csl'] = bool(int(col('CSL')))
+
+            if 'QMsgCnt' in headers:
+                child['qmsgcnt'] = int(col('QMsgCnt'))
```
Comment (abtink) [inline tests/scripts/thread-cert/node.py:1623]: Just curious, why we have the `if` checks here? Is it to check if `Ver` or `CSL` or `QMsgCnt` are included or not? I think they are always included in the table, right?

File: tests/scripts/thread-cert/test_inform_previous_parent_on_reattach.py (hunk 1 above)
Comment (simonlingoogle) [reply tests/scripts/thread-cert/test_inform_previous_parent_on_reattach.py:81]: `posix` does not have `INFORM_PREVIOUS_PARENT_ON_REATTACH` enabled by default.

File: tests/scripts/thread-cert/node.py (hunk 2 above)
Comment (simonlingoogle) [reply tests/scripts/thread-cert/node.py:1623]: > Just curious, why we have the if checks here? Is it to check if Ver or CSL or QMsgCnt are included or not?

Yes. 

>  I think they are always included in the table, right?

Yes. I copied the code from `otci.py`, which is compatible with older OT versions.  
Should be fined to remove here.

File: tests/scripts/thread-cert/test_inform_previous_parent_on_reattach.py (hunk 1 above)
Comment (abtink) [reply tests/scripts/thread-cert/test_inform_previous_parent_on_reattach.py:81]: Make sense. Thanks.

**Comment by simonlingoogle** [comment]
I created this test trying to reproduce a bug that `INFORM_PREVIOUS_PARENT_ON_REATTACH` feature was not working even when the `empty IPv6 frame` was properly sent. 

Latest tests seem to indicate that OTBR won't be able to remove the Child if TREL is enabled.

I haven't investigated what's the root cause yet, but wonder if you may have any clue. @abtink

File: tests/scripts/thread-cert/node.py (hunk 2 above)
Comment (simonlingoogle) [reply tests/scripts/thread-cert/node.py]: removed

**Comment by abtink** [comment]
> I created this test trying to reproduce a bug that INFORM_PREVIOUS_PARENT_ON_REATTACH feature was not working even when the empty IPv6 frame was properly sent.
> 
> Latest tests seem to indicate that OTBR won't be able to remove the Child if TREL is enabled.
> 
> I haven't investigated what's the root cause yet, but wonder if you may have any clue. @abtink

@simonlingoogle I have an idea but not 100% sure. 

I checked the code a bit.  The code on parent that checks this and kicks the child out when it sees the empty was added from https://github.com/openthread/openthread/pull/2182.
- The check is happening in `MeshForwarder::UpdateRoutes()` method
- This is called from `MeshForwarder::HandleMesh()` (if we received mesh header frame) or  from `HandleLowpanHC()` (if we receive a single frame), but I don't see it be called from `HandleFragment()`

One thing that is different when we have multi-radio we add frag header. I think this may be the cause of this that we don't go through this code path to kick the child out of the child table?

**Comment by simonlingoogle** [comment]
> One thing that is different when we have multi-radio we add frag header. I think this may be the cause of this that we don't go through this code path to kick the child out of the child table?

I think we can call `UpdateRoutes` for the first fragment or after the whole datagram is received. 

That being said, I am not quite familiar with the related code.

**Comment by abtink** [comment]
> I think we can call UpdateRoutes for the first fragment or after the whole datagram is received.

I was thinking the same. I submitted https://github.com/openthread/openthread/pull/7015 for this.
I think this may work (it will also update the address cache table for received fragmented msg). We can check if it addresses the issue.

**Review by jwhui — APPROVED** [review]

**Pull request merged**